File: levelledmobs-plugin-bukkit/src/main/java/me/lokka30/levelledmobs/bukkit/logic/nms/ComponentUtils.java
package me.lokka30.levelledmobs.bukkit.logic.nms;

import org.bukkit.entity.LivingEntity;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.InvocationTargetException;

public class ComponentUtils {
    public static void append(final @NotNull Object component, final @Nullable Object appendingComponent,
                              final @NotNull Definitions definitions) throws Exception {
        if (appendingComponent == null) return;

        if (component.getClass() != definitions.clazz_IChatMutableComponent){
            throw new Exception("Invalid type: " + component.getClass().getName());
        }
        if (appendingComponent.getClass() != definitions.clazz_IChatMutableComponent){
            throw new Exception("Invalid type: " + appendingComponent.getClass().getName());
        }

        definitions.method_ComponentAppend.invoke(component, appendingComponent);
    }

    public static @Nullable Object getEmptyComponent(final @NotNull Definitions definitions){
        return getTextComponent(null, definitions);
    }

    public static @Nullable Object getTextComponent(final @Nullable String text, final @NotNull Definitions definitions){
        try {
            if (text == null && definitions.getServerVersionInfo().getMinecraftVersion() >= 1.19) {
                // #empty()
                return definitions.method_EmptyComponent.invoke(null);
            } else {
                // #nullToEmpty(text)
                return definitions.method_TextComponent.invoke(null, text);
            }
        }
        catch (Exception e){
            e.printStackTrace();
        }

        return null;
    }

    public static @Nullable Object getTranslatableComponent(
            final @NotNull Definitions definitions, final @NotNull String key){
        return getTranslatableComponent(definitions, key, (Object) null);
    }

    public static @Nullable Object getTranslatableComponent(
            final @NotNull Definitions definitions, final @NotNull String key, Object... args){
        try {
            if (definitions.getServerVersionInfo().getMinecraftVersion() >= 1.19){
                if (args == null || args.length == 0) {
                    return definitions.method_Translatable.invoke(key);
                }
                else {
                    return definitions.method_TranslatableWithArgs.invoke(key, args);
                }
            }
            else{
                if (args == null || args.length == 0) {
                    return definitions.clazz_TranslatableComponent.getConstructor(String.class).newInstance(key);
                }
                else {
                    return definitions.clazz_TranslatableComponent.getConstructor(
                            String.class, Object[].class).newInstance(key, args);
                }
            }
        } catch (IllegalAccessException | InvocationTargetException | InstantiationException | NoSuchMethodException e) {
            e.printStackTrace();
        }

        return null;
    }

    public static @NotNull String getTranslationKey(final @NotNull LivingEntity livingEntity){
        // only needed for spigot. paper has a built-in method
        // TODO: implement the method below in reflection
        //return net.minecraft.world.entity.EntityType.byString(type.getName()).map(net.minecraft.world.entity.EntityType::getDescriptionId).orElse(null);
        return "";
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/nametag/ComponentUtils.kt
package io.github.arcaneplugins.levelledmobs.nametag

import io.github.arcaneplugins.levelledmobs.LevelledMobs

/**
 * Provides methods for manipulating internal text
 * components
 *
 * @author stumper66
 * @since 3.9.3
 */
object ComponentUtils {
    private val def = LevelledMobs.instance.definitions
    private val ver = LevelledMobs.instance.ver

    fun appendComponents(
        component: Any,
        appendingComponent: Any?
    ) {
        if (appendingComponent == null) {
            return
        }

        try {
            def.methodComponentAppend!!.invoke(component, appendingComponent)
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    fun getEmptyComponent(): Any {
        return getTextComponent(null)!!
    }

    fun getTextComponent(
        text: String?
    ): Any? {
        try {
            return def.methodTextComponent!!.invoke(null, text)
        } catch (e: Exception) {
            e.printStackTrace()
            return null
        }
    }

    fun getTranslatableComponent(
        key: String
    ): Any? {
        return getTranslatableComponent(key, null)
    }

    fun getTranslatableComponent(
        key: String,
        args: Array<Any?>?
    ): Any? {
        try {
            if (ver.minecraftVersion >= 1.19) {
                if (args.isNullOrEmpty() || args[0] == null) {
                    return def.methodTranslatable!!.invoke(null, key)
                }

                return def.methodTranslatableWithArgs!!.invoke(null, key, args)
            } else {
                return if (args.isNullOrEmpty()) {
                    def.clazzTranslatableComponent!!
                        .getConstructor(String::class.java)
                        .newInstance(key)
                } else {
                    def.clazzTranslatableComponent!!
                        .getConstructor(String::class.java, Array<Any>::class.java)
                        .newInstance(key, args)
                }
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
            return null
        }
    }
}